From ad4b5e68bb40f3e0bbf24cf4ad2949ae042e132e Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Mon, 30 Aug 2010 14:08:03 -0600
Subject: [PATCH] Refactor SettingsPanel to use a Listenable object

Create a ListenableAccountDiffPreference and use it from
the PatchScriptSettingsPanel so that the PatschScreen can
listen directly to the data model for changes instead of the
PatchScriptSettingsPanel.

Change-Id: I52fb415fcbc1ed15a453840ebde0de8b8d8f6710
---
 .../google/gerrit/client/patches/PatchScreen.java  |    9 +++-
 .../client/patches/PatchScriptSettingsPanel.java   |   40 ++++++++++---------
 .../client/ui/ListenableAccountDiffPreference.java |   16 ++++++++
 .../google/gerrit/client/ui/ListenableValue.java   |   34 +++++++++++++++++
 4 files changed, 77 insertions(+), 22 deletions(-)
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableValue.java

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
index 26b2853..789f1da 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
@@ -22,6 +22,7 @@ import com.google.gerrit.client.changes.PatchTable;
 import com.google.gerrit.client.changes.Util;
 import com.google.gerrit.client.rpc.GerritCallback;
 import com.google.gerrit.client.rpc.ScreenLoadCallback;
+import com.google.gerrit.client.ui.ListenableAccountDiffPreference;
 import com.google.gerrit.client.ui.Screen;
 import com.google.gerrit.common.data.PatchScript;
 import com.google.gerrit.common.data.PatchSetDetail;
@@ -166,14 +167,16 @@ public abstract class PatchScreen extends Screen implements
     idSideB = diffSideB != null ? diffSideB : id.getParentKey();
     this.patchIndex = patchIndex;
 
-    settingsPanel = new PatchScriptSettingsPanel();
-    settingsPanel
-        .addValueChangeHandler(new ValueChangeHandler<AccountDiffPreference>() {
+    ListenableAccountDiffPreference prefs =
+      new ListenableAccountDiffPreference();
+    prefs.addValueChangeHandler(new ValueChangeHandler<AccountDiffPreference>() {
           @Override
           public void onValueChange(ValueChangeEvent<AccountDiffPreference> event) {
             update(event.getValue());
           }
         });
+
+    settingsPanel = new PatchScriptSettingsPanel(prefs);
     settingsPanel.getReviewedCheckBox().addValueChangeHandler(
         new ValueChangeHandler<Boolean>() {
           @Override
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java
index 758b0f0..dade7f7 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java
@@ -17,6 +17,7 @@ package com.google.gerrit.client.patches;
 import com.google.gerrit.client.Gerrit;
 import com.google.gerrit.client.account.Util;
 import com.google.gerrit.client.rpc.GerritCallback;
+import com.google.gerrit.client.ui.ListenableAccountDiffPreference;
 import com.google.gerrit.client.ui.NpIntTextBox;
 import com.google.gerrit.reviewdb.AccountDiffPreference;
 import com.google.gerrit.reviewdb.AccountDiffPreference.Whitespace;
@@ -48,7 +49,7 @@ public class PatchScriptSettingsPanel extends Composite implements
   interface MyUiBinder extends UiBinder<Widget, PatchScriptSettingsPanel> {
   }
 
-  private AccountDiffPreference value;
+  private ListenableAccountDiffPreference listenablePrefs;
   private boolean enableIntralineDifference = true;
   private boolean enableSmallFileFeatures = true;
 
@@ -96,7 +97,8 @@ public class PatchScriptSettingsPanel extends Composite implements
    */
   private int setEnabledCounter;
 
-  public PatchScriptSettingsPanel() {
+  public PatchScriptSettingsPanel(ListenableAccountDiffPreference prefs) {
+    listenablePrefs = prefs;
     initWidget(uiBinder.createAndBindUi(this));
     initIgnoreWhitespace(ignoreWhitespace);
     initContext(context);
@@ -115,11 +117,7 @@ public class PatchScriptSettingsPanel extends Composite implements
     tabWidth.addKeyPressHandler(onEnter);
     colWidth.addKeyPressHandler(onEnter);
 
-    if (Gerrit.isSignedIn() && Gerrit.getAccountDiffPreference() != null) {
-      setValue(Gerrit.getAccountDiffPreference());
-    } else {
-      setValue(AccountDiffPreference.createDefault(null));
-    }
+    display();
   }
 
   @Override
@@ -147,7 +145,7 @@ public class PatchScriptSettingsPanel extends Composite implements
   public void setEnableSmallFileFeatures(final boolean on) {
     enableSmallFileFeatures = on;
     if (enableSmallFileFeatures) {
-      syntaxHighlighting.setValue(value.isSyntaxHighlighting());
+      syntaxHighlighting.setValue(getValue().isSyntaxHighlighting());
     } else {
       syntaxHighlighting.setValue(false);
     }
@@ -157,7 +155,7 @@ public class PatchScriptSettingsPanel extends Composite implements
   public void setEnableIntralineDifference(final boolean on) {
     enableIntralineDifference = on;
     if (enableIntralineDifference) {
-      intralineDifference.setValue(value.isIntralineDifference());
+      intralineDifference.setValue(getValue().isIntralineDifference());
     } else {
       intralineDifference.setValue(false);
     }
@@ -178,10 +176,16 @@ public class PatchScriptSettingsPanel extends Composite implements
   }
 
   public AccountDiffPreference getValue() {
-    return value;
+    return listenablePrefs.get();
   }
 
   public void setValue(final AccountDiffPreference dp) {
+    listenablePrefs.set(dp);
+    display();
+  }
+
+  protected void display() {
+    final AccountDiffPreference dp = getValue();
     setIgnoreWhitespace(dp.getIgnoreWhitespace());
     if (enableSmallFileFeatures) {
       syntaxHighlighting.setValue(dp.isSyntaxHighlighting());
@@ -195,8 +199,6 @@ public class PatchScriptSettingsPanel extends Composite implements
     intralineDifference.setValue(dp.isIntralineDifference());
     whitespaceErrors.setValue(dp.isShowWhitespaceErrors());
     showTabs.setValue(dp.isShowTabs());
-
-    value = dp;
   }
 
   @UiHandler("update")
@@ -205,7 +207,7 @@ public class PatchScriptSettingsPanel extends Composite implements
   }
 
   private void update() {
-    AccountDiffPreference dp = new AccountDiffPreference(value);
+    AccountDiffPreference dp = new AccountDiffPreference(getValue());
     dp.setIgnoreWhitespace(getIgnoreWhitespace());
     dp.setContext(getContext());
     dp.setTabSize(tabWidth.getIntValue());
@@ -215,8 +217,7 @@ public class PatchScriptSettingsPanel extends Composite implements
     dp.setShowWhitespaceErrors(whitespaceErrors.getValue());
     dp.setShowTabs(showTabs.getValue());
 
-    value = dp;
-    fireEvent(new ValueChangeEvent<AccountDiffPreference>(dp) {});
+    listenablePrefs.set(dp);
 
     if (Gerrit.isSignedIn()) {
       persistDiffPreferences();
@@ -225,10 +226,11 @@ public class PatchScriptSettingsPanel extends Composite implements
 
   private void persistDiffPreferences() {
     setEnabled(false);
-    Util.ACCOUNT_SVC.changeDiffPreferences(value, new GerritCallback<VoidResult>() {
+    Util.ACCOUNT_SVC.changeDiffPreferences(getValue(),
+        new GerritCallback<VoidResult>() {
       @Override
       public void onSuccess(VoidResult result) {
-        Gerrit.setAccountDiffPreference(value);
+        Gerrit.setAccountDiffPreference(getValue());
         setEnabled(true);
       }
 
@@ -267,7 +269,7 @@ public class PatchScriptSettingsPanel extends Composite implements
     if (0 <= sel) {
       return Whitespace.valueOf(ignoreWhitespace.getValue(sel));
     }
-    return value.getIgnoreWhitespace();
+    return getValue().getIgnoreWhitespace();
   }
 
   private void setIgnoreWhitespace(Whitespace s) {
@@ -285,7 +287,7 @@ public class PatchScriptSettingsPanel extends Composite implements
     if (0 <= sel) {
       return Short.parseShort(context.getValue(sel));
     }
-    return (short) value.getContext();
+    return (short) getValue().getContext();
   }
 
   private void setContext(int ctx) {
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java
new file mode 100644
index 0000000..11ee50e
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java
@@ -0,0 +1,16 @@
+package com.google.gerrit.client.ui;
+
+import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.reviewdb.AccountDiffPreference;
+
+public class ListenableAccountDiffPreference
+    extends ListenableValue<AccountDiffPreference> {
+
+  public ListenableAccountDiffPreference() {
+    if (Gerrit.isSignedIn() && Gerrit.getAccountDiffPreference() != null) {
+      set(Gerrit.getAccountDiffPreference());
+    } else {
+      set(AccountDiffPreference.createDefault(null));
+    }
+  }
+}
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableValue.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableValue.java
new file mode 100644
index 0000000..4fc4d6f
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableValue.java
@@ -0,0 +1,34 @@
+package com.google.gerrit.client.ui;
+
+import com.google.gwt.event.logical.shared.HasValueChangeHandlers;
+import com.google.gwt.event.logical.shared.ValueChangeEvent;
+import com.google.gwt.event.logical.shared.ValueChangeHandler;
+import com.google.gwt.event.shared.GwtEvent;
+import com.google.gwt.event.shared.HandlerManager;
+import com.google.gwt.event.shared.HandlerRegistration;
+
+
+public class ListenableValue<T> implements HasValueChangeHandlers<T> {
+
+  private HandlerManager manager = new HandlerManager(this);
+
+  private T value;
+
+  public T get() {
+    return value;
+  }
+
+  public void set(final T value) {
+    this.value = value;
+    fireEvent(new ValueChangeEvent<T>(value) {});
+  }
+
+  public void fireEvent(GwtEvent<?> event) {
+    manager.fireEvent(event);
+  }
+
+  public HandlerRegistration addValueChangeHandler(
+      ValueChangeHandler<T> handler) {
+    return manager.addHandler(ValueChangeEvent.getType(), handler);
+  }
+}
-- 
1.7.2.2

